core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowRewriter.scala

```diff
@@ -0,0 +1,424 @@
+    var cacheSinkOperatorDescriptors: mutable.HashMap[String, CacheSinkOpDesc],
+    var operatorRecord: mutable.HashMap[String, WorkflowVertex],
+    var opResultStorage: OpResultStorage
+) {
+
+  private val logger = Logger(this.getClass.getName)
+
+  var visitedOpID: mutable.HashSet[String] = new mutable.HashSet[String]()
+
+  private val workflowDAG: WorkflowDAG = if (workflowInfo != null) {
+    new WorkflowDAG(workflowInfo)
+  } else {
+    null
+  }
+  private val rewrittenToCacheOperatorIDs = if (null != workflowInfo) {
+    new mutable.HashSet[String]()
+  } else {
+    null
+  }
+  private val r1NewOps = if (workflowInfo != null) {
+    mutable.MutableList[OperatorDescriptor]()
+  } else {
+    null
+  }
+  private val r1NewLinks = if (workflowInfo != null) {
+    mutable.MutableList[OperatorLink]()
+  } else {
+    null
+  }
+  private val r1NewBreakpoints = if (workflowInfo != null) {
+    mutable.MutableList[BreakpointInfo]()
+  } else {
+    null
+  }
+  private val r2NewOps = if (workflowInfo != null) {
+    mutable.MutableList[OperatorDescriptor]()
+  } else {
+    null
+  }
+  private val r2NewLinks = if (workflowInfo != null) {
+    mutable.MutableList[OperatorLink]()
+  } else {
+    null
+  }
+  private val r2NewBreakpoints = if (workflowInfo != null) {
+    mutable.MutableList[BreakpointInfo]()
+  } else {
+    null
+  }
+  private val r2OpIDQue = if (workflowInfo != null) {
+    new mutable.Queue[String]()
+  } else {
+    null
+  }
+
+  var r1WorkflowInfo: WorkflowInfo = _
+  var r1WorkflowDAG: WorkflowDAG = _
+  var r2WorkflowInfo: WorkflowInfo = _
+  var r2WorkflowDAG: WorkflowDAG = _
+
+  def rewrite: WorkflowInfo = {
+    if (null == workflowInfo) {
+      logger.info("Rewriting workflow null")
+      null
+    } else {
+      logger.info("Rewriting workflow {}", workflowInfo)
+      checkCacheValidity()
+
+      workflowDAG.getSinkOperators.foreach(r2OpIDQue.+=)
+      while (r2OpIDQue.nonEmpty) {
+        r2(r2OpIDQue.dequeue())
+      }
+      val r2TmpLinks = mutable.MutableList[OperatorLink]()
+      val r2TmpOpIDs = r2NewOps.map(op => op.operatorID)
+      r2NewLinks.foreach(link => {
+        if (r2TmpOpIDs.contains(link.destination.operatorID)) {
+          if (r2TmpOpIDs.contains(link.origin.operatorID)) {
+            r2TmpLinks += link
+          }
+        }
+      })
+      r2NewLinks.clear()
+      r2TmpLinks.foreach(r2NewLinks.+=)
+
+      r2WorkflowInfo = WorkflowInfo(r2NewOps, r2NewLinks, r2NewBreakpoints)
+      r2WorkflowDAG = new WorkflowDAG(r2WorkflowInfo)
+      r2WorkflowDAG.getSinkOperators.foreach(r2OpIDQue.+=)
+
+      val r1OpIDIter = r2WorkflowDAG.jgraphtDag.iterator()
+      var r1OpIDs: mutable.MutableList[String] = mutable.MutableList[String]()
+      r1OpIDIter.forEachRemaining(id => r1OpIDs.+=(id))
+      r1OpIDs = r1OpIDs.reverse
+      r1OpIDs.foreach(r1)
+
+      new WorkflowInfo(r1NewOps, r1NewLinks, r1NewBreakpoints)
+    }
+  }
+
+  private def r1(opID: String): Unit = {
```

Comment:
this function name is also very much unreadable. please consider change it or add enough comments.